initial document for benchmark (#219)
dragonflydb/documentation: docs/getting-started/benchmark.md

For instance, to minimize the impact of environment and network latency, we strongly recommend
running both machines as close together as possible, ideally within the same
Availability Zone (AZ) and region. For the AWS cloud, consider using a "Cluster" placement group
to achieve the lowest possible latency.

The remainder of this document will discuss how to set up a benchmark in the AWS cloud
to observe millions of QPS from a single instance.

## Load testing c6gn.12xlarge
I usually choose a loadtest machine to be stronger than the server instance to eliminate
any possible bottlenecks on the client side. For this setup, I used `c7gn.16xlarge` for running
the loadtest client and `c6gn.12xlarge` for running dragonfly.
I used Dragonfly v1.15.0 (latest at the time of writing) with the following arguments:
`./dragonfly --logtostderr  --dbfilename=`

Please notice that Dragonfly will uses all the available vCPUs by default, in this case 48 vCPUs
If you want to control explicitly number of threads in Dragonfly you can add `--proactor_threads=<N>`
argument.
Both machines run `Ubuntu 23.04` OS with kernel version 6.2.

### Writes-only test

> On the loadtest instance (c7gn.16xlarge with 64 vCPUs) I run:

Reviewer: Although, thinking about this a bit more, consider using present tense in the form of: "Let's try to run Dragonfly in X mode: ...".
Your call.

> `memtier_benchmark -s $SERVER_IP --distinct-client-seed --hide-histogram --ratio 1:0 -t 60 -c 20 -n 200000`

Author: good catch

The run ended with the following summary:

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets      4195628.23          ---          ---         0.39283         0.37500         0.68700         2.54300    323231.06

```

### Reads-only test

> Without flushing the database, I run the following command:

Reviewer: ```suggestion
Without flushing the database, I ran the following command:
```

`memtier_benchmark -s $SERVER_IP --distinct-client-seed --hide-histogram --ratio 0:1 -t 60 -c 20 -n 200000`

> notice the ratio changed to "1:0.

Reviewer: ```suggestion
notice the ratio changed to "0:1", meaning only `SET` commands and *no* `GET` commands.
```

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      4109802.84   4109802.84         0.00         0.40126         0.38300         0.67900         0.90300    296551.68
```

> 

Reviewer: Again, I'd add what these numbers mean

### Read test with pipelining

> There are many other ways to loadtest Dragonfly. Here is one with sending SETs with pipeline of batch size 10:

Reviewer: ```suggestion
There are many other ways to loadtest Dragonfly. Below is one with sending `SET`s with pipeline (`--pipeline`) of batch size 10. Pipeline means that the client sends multiple commands (10 in this case) and only then waits for the responses.
```

`memtier_benchmark -s $SERVER_IP --ratio 0:1 -t 60 -c 5  -n 200000  --distinct-client-seed --hide-histogram --pipeline=10`

```
ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      7083583.57   7083583.57         0.00         0.45821         0.44700         0.69500         1.53500    511131.14
```

> ## Load testing c7gn.12xlarge

Reviewer: ```suggestion
## Load testing `c7gn.12xlarge`
```

> Next thing I tried running Dragonfly on the next generation instance with the same number of vCPUs (48).

Reviewer: ```suggestion
Next, I tried running Dragonfly on the next generation instance (`c7g` vs `c6g`) with the same number of vCPUs (48).
```

---
sidebar_position: 3
---

# Benchmarking Dragonfly

Dragonfly is a high-performance, distributed key-value store designed for scalability
and low latency. It's a drop-in replacement for Redis 6.x and memcached servers.
This document outlines a benchmarking methodology and results achieved using Dragonfly
with the `memtier_benchmark` load testing tool.

We benchmarked Dragonfly using the [memtier_benchmark](https://github.com/RedisLabs/memtier_benchmark)
load testing tool.
A prebuilt container is also available on [Docker Hub](https://hub.docker.com/r/redislabs/memtier_benchmark/)
Although Redis offers the `redis-benchmark` tool in its repository, it has not been as efficient
as `memtier_benchmark` and it often becomes the bottleneck instead of Dragonfly.

We also developed our own tool [dfly_bench](https://github.com/dragonflydb/dragonfly/blob/main/src/server/dfly_bench.cc),
which can be built from source in the Dragonfly repository.

## Methodology
- **Remote Deployment:** Dragonfly is a multi-threaded server designed to run remotely.

> Therefore, we recomment running the load testing client and server on separate machines for a more accurate representation of real-world performance.

Reviewer: ```suggestion
Therefore, we recommend running the load testing client and server on separate machines for a more accurate representation of real-world performance.
```

- **Minimizing Latency:** Locate client and server within the same Availability Zone and use private IPs for optimal network performance. If you benchmark in the AWS cloud, consider an AWS Cluster placement group
for the lowest possible latency. The rationale behind this - to remove any environmental factors
that might skew the test results.
- **Server vs. Client Resources**: Use a more powerful instance for the load testing client
to avoid client-side bottlenecks.

The remainder of this document will discuss how to set up a benchmark in the AWS cloud
to observe millions of QPS from a single instance.

## Load testing configuration
We used Dragonfly v1.15.0 (latest at the time of writing) with the following arguments:
`./dragonfly --logtostderr  --dbfilename=`

Please notice that Dragonfly uses of all the available vCPUs by default on the server.
If you want to control explicitly number of threads in Dragonfly you can add `--proactor_threads=<N>`.
Both the client and server instances run `Ubuntu 23.04` OS with kernel version 6.2.
In line with our recommendations above, we used internal IPs for connecting and

> used stronger `c7gn.16xlarge` instance with 64 vCPUs for the load-testing program.

Reviewer: ```suggestion
used stronger `c7gn.16xlarge` instance with 64 vCPUs for the load-testing program (i.e. the client).
```

## Dragongly on `c6gn.12xlarge`

### Write-only test

> On the loadtest instance (c7gn.16xlarge with 64 vCPUs) we run:

Reviewer: ```suggestion
On the loadtest instance (`c7gn.16xlarge` with 64 vCPUs):
```

Consider removing the "we run" part (here and probably below)

`memtier_benchmark -s $SERVER_PRIVATE_IP --distinct-client-seed --hide-histogram --ratio 1:0 -t 60 -c 20 -n 200000`

The run ended with the following summary:

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets      4195628.23          ---          ---         0.39283         0.37500         0.68700         2.54300    323231.06

```

In this test, we reached almost 4.2M queries per second (QPS) with the average latency of

> 392us between the `memtier_benchmark` and Dragonfly. Consequently, the P50 latency was 375us, P99 - 687us

Reviewer: ```suggestion
392us (0.4ms) between the `memtier_benchmark` and Dragonfly. Consequently, the P50 latency was 375us (0.4ms), P99 - 687us (0.7ms)
```
It might be too much to add in all places, but at least for the first one it emphasizes the extreme low latency (and maybe also reduces confusion - "hundreds" sound like a lot)

and P99.9 was 2543us. It is a very short and simple test, but it still gives some perspective
about the performance of Dragonfly.

### Reads-only test
Without flushing the database, we run the following command:
`memtier_benchmark -s $SERVER_PRIVATE_IP --distinct-client-seed --hide-histogram --ratio 0:1 -t 60 -c 20 -n 200000`

> Note that the ratio changed to "1:0.

Reviewer: ```suggestion
Note that the `SET`/`GET` ratio (`--ratio`) changed to "1:0", meaning only `GET` operations.
```

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      4109802.84   4109802.84         0.00         0.40126         0.38300         0.67900         0.90300    296551.68
```

We can observe that both `Ops` and `Hits` are the same, meaning all of the GET requests
coming from the load test hit the existing keys.
Dragonfly responded with returning values for each request and its average QPS was 4.1M qps,

> with P99.9 latency - 903us.

Reviewer: ```suggestion
with P99.9 latency - 903us (less than 1 millisecond).
```

### Read test with pipelining

> Here's another way to loadtest Dragonfly. This one is sending SETs with pipeline of batch size 10:

Reviewer: ```suggestion
Here's another way to loadtest Dragonfly. This one is sending `SET`s with pipeline of batch size 10 (note the `--pipeline=10` flag):
```

`memtier_benchmark -s $SERVER_PRIVATE_IP --ratio 0:1 -t 60 -c 5  -n 200000  --distinct-client-seed --hide-histogram --pipeline=10`

```
ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      7083583.57   7083583.57         0.00         0.45821         0.44700         0.69500         1.53500    511131.14
```

> During the pipelining mode, memtier_benchmark sends `K` requests in batch without before waiting
> for them to complete. In this case, `K` is `10`. Pipelining reduces the CPU load spent

Reviewer: ```suggestion
During the pipelining mode, memtier_benchmark sends `K` (in this case 10) requests in batch without waiting
for them to complete. Pipelining reduces the CPU load spent
```

in the networking stack, and as a result, Dragonfly can reach 7M qps with sub-millisecond latency.
Please note, that for real world usecases, pipelining requires cooperation of a client side app,
which must send multiple requests on a single connection before waiting for the server to respond.

Some asynchronous client libraries like `StackExchange.Redis` or `ioredis` allow multiplexing requests
on a single connection. They can still provide a simplified synchronous interface to their users while
benefitting from performance improvements of pipelining.

## Load testing Dragonfly `c7gn.12xlarge`

Next thing we tried running Dragonfly on the next generation instance with the same number of vCPUs (48).
We used the same `c7gn.16xlarge` for running `memtier_benchmark` and we used the same commands
to test writes, reads and pipelined reads:

| Test          | Ops/sec   | Avg. Latency (us) | P99.9 Latency (us) |
|---------------|-----------|-------------------|--------------------|
| Write-Only    | 5.2M      | 250               | 631                |
| Read-Only     | 6M        | 271               | 623                |

> | Pipelined Read| 8.9M      | 323               | 839                |

Reviewer: Having a real table is much more readable, maybe use tables instead of console output above?

Author: I prefer to keep it for the first experiment to show the authenticity of the results.

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      6078632.89   6078632.89         0.00         0.27177         0.26300         0.49500         0.62300    438616.86
```

### Pipelined Reads
`memtier_benchmark -s $SERVER_PRIVATE_IP --ratio 0:1 -t 60 -c 5  -n 200000  --distinct-client-seed --hide-histogram --pipeline=10`

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      8975121.86   8975121.86         0.00         0.32325         0.31100         0.52700         0.83900    647619.14
```

## Comparison with Garnet

> Microsoft Research recently released Garnet ([https://github.com/microsoft/garnet]),

Reviewer: ```suggestion
Microsoft Research recently released [Garnet](https://github.com/microsoft/garnet),
```

a remote cache store. Due to interest within the Dragonfly community, we decided to compare
Garnet's performance with Dragonfly's. This comparison focuses on performance results
and does not delve into architectural differences or Redis compatibility implications.

> <i>Note: Unfortunately, Garnet does not have aarch64 build available,

Reviewer: I think that the usage of `<i>` here messes up the formatting. Maybe use `_` instead, or only make the `Note:` part italic?

therefore we run both Garnet and Dragonfly on x86_64 server
`c6in.12xlarge`. We run Garnet via docker with host networking enabled via
`docker run --network=host ghcr.io/romange/garnet:latest --port=6379` command.
The docker container was built using the Garnet docker build file for ubuntu, located in their
repository.</i>

### Garnet on `c6in.12xlarge`

Similarly to previous tests we run `memtier_benchmark` on `c7gn.16xlarge` with `cluster`
placement policy for both instances. For writes we used the following command:

```
memtier_benchmark -s $SERVER_PRIVATE_IP --distinct-client-seed --hide-histogram --ratio 1:0 -t 60 -c 20 -n 200000
```

Similarly, for reads we used
```
memtier_benchmark -s $SERVER_PRIVATE_IP --distinct-client-seed --hide-histogram --ratio 0:1 -t 60 -c 20 -n 200000
```

and for pipelined reads we used

```
memtier_benchmark -s $SERVER_PRIVATE_IP --ratio 0:1 -t 60 -c 5  -n 2000000  --distinct-client-seed --hide-histogram --pipeline=10
```

> Note that we increased number of requests to `2000000` per client connection in the latter case.

Reviewer: why though? and why not in the above tests?

Author: the pipelining QPS was so high that `200000` finishes too quickly

```

### Pipelined Reads
`memtier_benchmark -s $SERVER_PRIVATE_IP --ratio 0:1 -t 60 -c 5  -n 200000  --distinct-client-seed --hide-histogram --pipeline=10`

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Gets      8975121.86   8975121.86         0.00         0.32325         0.31100         0.52700         0.83900    647619.14
```

## Garnet on `c6in.12xlarge`
[Garnet](https://github.com/microsoft/garnet) has been recently released by Microsoft Research
and people are curious how it compares with Dragonfly. We are not going to cover here about
the architectural differrences between Garnet and Dragonfly, and their implications on compatibility
with Redis but we provide a brief summary of Garnet performance and compare it with Dragonfly.

> Unfortunately, Garnet does not have aarch64 build available so we run it on intel-powered server

Reviewer: Consider doing the whole page using x64, as otherwise it's just incomparable and users will either be confused or angry that they are "being mislead" (even though that is clearly not the goal here)

Author: I added Dragonfly results below for the same I stance, I have not finished.
